File: Android/src/main/java/com/cisco/or/sdk/utils/SharedPrefs.kt
package com.cisco.or.sdk.utils

import android.content.Context
import android.content.SharedPreferences
import com.cisco.or.sdk.enums.SigningState
import java.security.Key
import java.security.KeyPair
import java.security.PublicKey

internal class SharedPrefs(context:Context) {

    companion object{
        private const val PREF_SDK_TOKEN = "PREF_SDK_TOKEN"
        private const val PREF_SIGNING_STATE = "PREF_SIGNING_STATE"
        private const val FQDN = "INSTALLED_FQDN"
        private const val PUSH_IDENTIFIER = "PREF_PUSH_IDENTIFIER"
        private const val PREF_DNA_SPACES_KEY = "PREF_DNA_SPACES_KEY"
        private const val CLIENT_PUBLIC_KEY = "PREF_CLIENT_PUBLIC_KEY"
        private const val CLIENT_PRIVATE_KEY = "PREF_CLIENT_PRIVATE_KEY"
        private const val APP_ID = "APP_ID"
    }

    private val sharedPref: SharedPreferences = context.getSharedPreferences(Constants.PREF_NAME, Context.MODE_PRIVATE)

    var clientPublicKey:String?
        get() = sharedPref.getString(CLIENT_PUBLIC_KEY, null)
        set(value) = sharedPref.edit().putString(CLIENT_PUBLIC_KEY, value).apply()
    var clientPrivateKey:String?
        get() = sharedPref.getString(CLIENT_PRIVATE_KEY, null)
        set(value) = sharedPref.edit().putString(CLIENT_PRIVATE_KEY, value).apply()

    var appId:String?
        get() = sharedPref.getString(APP_ID, null)
        set(value) = sharedPref.edit().putString(APP_ID, value).apply()

    var dnaSpacesKey:String?
        get() = sharedPref.getString(PREF_DNA_SPACES_KEY, null)
        set(value) = sharedPref.edit().putString(PREF_DNA_SPACES_KEY, value).apply()

    var sdkToken:String?
        get() = sharedPref.getString(PREF_SDK_TOKEN, null)
        set(value) = sharedPref.edit().putString(PREF_SDK_TOKEN, value).apply()

    var signingState:SigningState
    get() = when (sharedPref.getInt(PREF_SIGNING_STATE, SigningState.UNSIGNED.ordinal)){
        SigningState.SIGNED.ordinal -> SigningState.SIGNED
        else -> SigningState.UNSIGNED
    }
    set(value) = sharedPref.edit().putInt(PREF_SIGNING_STATE, value.ordinal).apply()

    var fqdn:String?
        get() = sharedPref.getString(FQDN, "")
        set(value) = sharedPref.edit().putString(FQDN, value).apply()


    var pushIdentifier:String?
        get() = sharedPref.getString(PUSH_IDENTIFIER, "")
        set(value) = sharedPref.edit().putString(PUSH_IDENTIFIER, value).apply()

}

File: IOS/Sources/OpenRoaming/OpenRoaming/Utils/SharedPrefs.swift
//
//  SharedPrefs.swift
//  OpenRoaming
//
//  Created by Luis Vasquez on 10/04/20.
//  Copyright © 2020 Cisco. All rights reserved.
//

import UIKit

class SharedPrefs: NSObject {
    
    private let PREF_APP_ID = "PREF_CLIENT_ID"
    private let PREF_CLIENT_PUBLIC_KEY = "PREF_CLIENT_PUBLIC_KEY"
    private let PREF_CLIENT_PRIVATE_KEY = "PREF_CLIENT_PRIVATE_KEY"
    private let DNA_SPACES_KEY = "DNA_SPACES_KEY"
    private let PREF_ACCESS_TOKEN = "PREF_ACCESS_TOKEN"
    private let PREF_SDK_TOKEN = "PREF_SDK_TOKEN"
    private let PREF_TERMS = "PREF_TERMS"
    private let PREF_SIGNING_STATE = "PREF_SIGNING_STATE"
    private let PREF_SIGNING_SERVICE_NAME = "PREF_SIGNING_SERVICE_NAME"
    private let DOMAIN_NAME = "DomainName"
    private let FQDN = "INSTALLED_FQDN"
    private let PREF_PUSH_ID = "PREF_PUSH_ID"
    
    var domainName: String? {
        get {
            UserDefaults.standard.string(forKey: DOMAIN_NAME)
        }
        set(value){
            UserDefaults.standard.set(value, forKey: DOMAIN_NAME)
        }
    }
    
    var appId: String? {
        get {
            UserDefaults.standard.string(forKey: PREF_APP_ID)
        }
        set(value) {
            UserDefaults.standard.set(value, forKey: PREF_APP_ID)
        }
    }
    
    var clientPublicKey: String? {
        get {
            UserDefaults.standard.string(forKey: PREF_CLIENT_PUBLIC_KEY)
        }
        set(value) {
            UserDefaults.standard.set(value, forKey: PREF_CLIENT_PUBLIC_KEY)
        }
    }
    
    var clientPrivateKey: String? {
        get {
            UserDefaults.standard.string(forKey: PREF_CLIENT_PRIVATE_KEY)
        }
        set(value) {
            UserDefaults.standard.set(value, forKey: PREF_CLIENT_PRIVATE_KEY)
        }
    }
    
    var dnaSpacesKey: String? {
        get {
            UserDefaults.standard.string(forKey: DNA_SPACES_KEY)
        }
        set(value) {
            UserDefaults.standard.set(value, forKey: DNA_SPACES_KEY)
        }
    }
    
    var accessToken: String? {
        get {
            UserDefaults.standard.string(forKey: PREF_ACCESS_TOKEN)
        }
        set(value) {
            UserDefaults.standard.set(value, forKey: PREF_ACCESS_TOKEN)
        }
    }
    
    var sdkToken: String? {
        get {
            UserDefaults.standard.string(forKey: PREF_SDK_TOKEN)
        }
        set(value) {
            UserDefaults.standard.set(value, forKey: PREF_SDK_TOKEN)
        }
    }
    
    var acceptTerms: Bool {
        get {
            UserDefaults.standard.bool(forKey: PREF_TERMS)
        }
        set(value) {
            UserDefaults.standard.set(value, forKey: PREF_TERMS)
        }
    }
    
    var signingState: SigningState {
        get {
            let i = UserDefaults.standard.string(forKey: PREF_SIGNING_STATE)
            switch(i){
                case SigningState.signed.rawValue:
                    return SigningState.signed
                default:
                    return SigningState.unsigned
            }
        }
        set(value) {
            UserDefaults.standard.set(value.rawValue, forKey: PREF_SIGNING_STATE)
        }
    }
    
    var signingServiceName: String {
        get {
            UserDefaults.standard.string(forKey: PREF_SIGNING_SERVICE_NAME) ?? ""
        }
        set(value) {
            UserDefaults.standard.set(value, forKey: PREF_SIGNING_SERVICE_NAME)
        }
    }
    
    var fqdn: String {
        get {
            UserDefaults.standard.string(forKey: FQDN) ?? ""
        }
        set(value) {
            UserDefaults.standard.set(value, forKey: FQDN)
        }
    }
    
    var pushIdentifier: String {
        get {
            UserDefaults.standard.string(forKey: PREF_PUSH_ID) ?? ""
        }
        set(value) {
            UserDefaults.standard.set(value, forKey: PREF_PUSH_ID)
        }
    }
    
    func logout() {
        accessToken = nil
        signingState = SigningState.unsigned
    }
}
